### PR TITLE
Add capability to validate FOTA image type before installing

### PR DESCRIPTION
**Pull request by jayteemo**
Specify the expected image type when starting a FOTA download.
If the actual type does not match, return failure before installation is started.

### CODE REVIEW

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c [hunk 1]
```diff
@@ -856,9 +860,24 @@ static int start_job(struct nrf_cloud_fota_job *const job)
...
+
+	switch (job->info.type) {
+	case NRF_CLOUD_FOTA_BOOTLOADER:
+	case NRF_CLOUD_FOTA_APPLICATION:
+		img_type = DFU_TARGET_IMAGE_TYPE_MCUBOOT;
+		break;
+	case NRF_CLOUD_FOTA_MODEM:
+		img_type = DFU_TARGET_IMAGE_TYPE_MODEM_DELTA;
+	default:
+		LOG_ERR("Unhandled FOTA type: %d", job->info.type);
```
Comment (plskeggs) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:873]: Would it make sense to support non-standard types by passing unrecognized ones to the original `fota_download_start`?

**Review by plskeggs — APPROVED** [review]
LGTM

**Review by rlubos — APPROVED** [review]

**Comment by junqingzou** [comment]
Looks good, and wait to call new fota_download_start_with_validation(). BTW, the API that I added fota_download_target() became redundant now since we now don't do blind downloading like before.

**Review by junqingzou — APPROVED** [review]

**Review by jtguggedal — APPROVED** [review]

File: include/dfu/dfu_target.h [hunk 2]
```diff
@@ -21,6 +21,7 @@ extern "C" {
 #endif
 
 enum dfu_target_image_type {
+	DFU_TARGET_IMAGE_TYPE_INVALID = 0,
```
Comment (hakonfam) [inline include/dfu/dfu_target.h:24]: IMO this should be called `DFU_TARGET_IMAGE_TYPE_ANY` as that is how it is used. That would also improve readability in the code. If we need a specific type for `INVALID` later we should add that in addition to this.

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: include/dfu/dfu_target.h (hunk 2 above)
Comment (jayteemo) [reply include/dfu/dfu_target.h:24]: updated.

File: include/net/fota_download.h [hunk 3]
```diff
@@ -107,6 +110,29 @@ int fota_download_init(fota_download_callback_t client_callback);
...
+ * @param apn Access Point Name to use or NULL to use the default APN.
+ * @param fragment_size Fragment size to be used for the download.
+ *			If 0, @option{CONFIG_DOWNLOAD_CLIENT_HTTP_FRAG_SIZE} is used.
+ * @param expected_type Type of firmware file to be downloaded and installed.
+ *
+ * @retval 0	     If download has started successfully.
+ * @retval -EALREADY If download is already ongoing.
+ *                   Otherwise, a negative value is returned.
+ */
+int fota_download_start_with_validation(const char *host, const char *file,
```
Comment (hakonfam) [inline include/net/fota_download.h:132]: The name of this function is misleading, there is no validation performed of the image downloaded, its just checking the type of the image. What about `fota_download_start_type` or `fota_download_start_with_type` ?

**Review by hakonfam — CHANGES_REQUESTED** [review]
Comment was not included in last round, sorry about that :)

File: include/net/fota_download.h (hunk 3 above)
Comment (jayteemo) [reply include/net/fota_download.h]: updated

**Review by hakonfam — APPROVED** [review]

**Pull request merged**